### PR TITLE
net: restore TcpStream::{poll_read_ready, poll_write_ready}

### LINKED ISSUE #2228: Detect TCP disconnect without reading or writing anything to that socket?
## Version

* tokio: v0.2.11
  * tokio::net::tcp::TcpStream

## Description

In #483, it was discussed to use `poll_read_ready(Ready::empty())` for detecting TCP disconnect.

After that, `poll_read_ready()` was removed at 6d8cc4e4755abbd0baac9abf154837b9be011a07 according to a discussion on gitter as described in https://github.com/tokio-rs/tokio/pull/1392#issuecomment-519533786.

Is there any alternative way to detect TCP disconnect without writing or reading anything to/from that socket?

I haven't tried yet, but we might be able to use `peek()` or `poll_peek()` for this purpose.

### Background

In some situations, it's necessary to detect TCP disconnect without reading or writing anything to/from that socket.

For example, suppose an HTTP server that performs a time-consuming task for a response.  This HTTP server don't like to read data from the TCP socket for detecting disconnect in order to support HTTP keep-alive and request pipelining easily.  This HTTP server has no data to write until the task outputs something.

If there is no way to detect TCP disconnect other than reading/writing, this HTTP server cannot detect TCP disconnect until the task outputs something.

The similar situation occurs in `actix-web`.  And it seems to leak TCP sockets if clients disconnect while it waits for data from a `Stream` object:
https://github.com/masnagam/rust-case-studies/tree/master/actix-web-streaming-and-tcp-disconnect

### Related Issues

* #29
* #483
* #1182

### Example of detecting TCP disconnect using `poll_read_ready()`

* https://github.com/vi/websocat/commit/d7b0db2f8d12e6aa21dc5f1e3c6a0f04b905dd25

### PR DESCRIPTION
**Pull request by masnagam**
## Motivation

In some situations, it's necessary to detect TCP disconnect without reading or writing anything to/from that socket.

This PR restores `TcpStream::{poll_read_ready, poll_write_ready}` which were removed at 6d8cc4e4755abbd0baac9abf154837b9be011a07.

## Solution

Restore `TcpStream::{poll_read_ready, poll_write_ready}`.

New test cases for those functions are added.  See [tcp_ready.rs](https://github.com/tokio-rs/tokio/compare/master...masnagam:issue-2228?expand=1#diff-21b8b0d4875ff6f6dc4385f4d5b8f07f).

This PR closes #2228.

### CODE REVIEW

**Comment by carllerche** [comment]
cc @seanmonstar

**Comment by seanmonstar** [comment]
Thanks for the PR!

As described in #2728, we want to remove the public dependency on `mio` from the API. That issue describes a proposal around how to expose readiness for socket types. The internals are changing in #2828. Once that lands, would you be interested in trying to implement what's proposed in #2728?

**Comment by masnagam** [comment]
@seanmonstar I created [a new working branch](https://github.com/masnagam/tokio/tree/issue-2228-wip) for that.

I implemented [TcpStream::is_readable](https://github.com/masnagam/tokio/blob/issue-2228-wip/tokio/src/net/tcp/stream.rs#L260) and [TcpStream::is_writable](https://github.com/masnagam/tokio/blob/issue-2228-wip/tokio/src/net/tcp/stream.rs#L300) for testing purposes (I'll implement **correct** methods later).  These methods simply call `PollEvented::poll_xxx_ready` in `poll_fn` like what `TcpStream::connect_addr` do.

Unfortunately, [tests](https://github.com/masnagam/tokio/blob/issue-2228-wip/tokio/tests/tcp_ready.rs) fail, which passed before the rebase.

Probably, there are some mistakes in my changes.
Could you point out my mistakes in my changes if you can found?

**Comment by masnagam** [comment]
#3130 was merged.  Should we close this PR, or continue implementing `poll_[read|write]_ready` for `poll_[read|write]`?

**Comment by Darksonn** [comment]
We might just have an `poll_ready` that takes an `Interest`? What do you think @carllerche?

**Comment by carllerche** [comment]
I don't think we can have `poll_ready`. `poll_read_ready` and `poll_write_ready` makes it easier to explain the contract. There can only be one task waiting on each fn. The async fn version doesn't have this limitation because of how it is implemented.

File: tokio/src/net/tcp/stream.rs [hunk 1]
```diff
@@ -219,6 +219,91 @@ impl TcpStream {
...
+    ///         stream.poll_read_ready(cx, mio::Ready::readable())
+    ///     }).await?;
+    ///
+    ///     Ok(())
+    /// }
+    /// ```
+    pub fn poll_read_ready(
+        &self,
+        cx: &mut Context<'_>,
+        mask: mio::Ready,
```
Comment (carllerche) [inline tokio/src/net/tcp/stream.rs:264]: I would remove the mask.

**Comment by carllerche** [comment]
I think the various socket types can still provide `poll_read_ready` and `poll_write_ready` even though there are `async fn` variants. Not everyone can use those (i.e. future impls).

**Comment by masnagam** [comment]
PR has been rebased onto the current master branch.

`TcpStream::poll_[read|write]_ready` have been reimplemented using `Registration::poll_[read|write]_ready`.

File: tokio/src/net/tcp/stream.rs (hunk 1 above)
Comment (masnagam) [reply tokio/src/net/tcp/stream.rs]: @carllerche PR has been rebased on to the current master branch. This method has been reimplemented using `Registration::poll_read_ready`.

**Review by carllerche — APPROVED** [review]
Thanks :+1:

**Pull request merged**